src/commands/reword.rs

```diff
@@ -0,0 +1,314 @@
+use crate::opts::{MoveOptions, RebaseOptions};
+
+/// Reword a commit and restack it's descendants.
+#[instrument]
+pub fn reword(
+    effects: &Effects,
+    force: bool,
+    hashes: Vec<String>,
+    messages: Vec<String>,
+    git_run_info: &GitRunInfo,
+    rebase_options: &RebaseOptions,
+) -> eyre::Result<isize> {
+    let (repo, mut dag) = get_repo_and_dag(effects)?;
+
+    let commits = match resolve_commits_from_hashes(&repo, &mut dag, effects, hashes)? {
+        Some(commits) => commits,
+        None => return Ok(1),
+    };
+
+    let messages = match build_messages(&messages, &commits)? {
+        BuildRewordMessageResult::Succeeded { messages } => messages,
+        BuildRewordMessageResult::IdenticalMessage => {
+            println!("The message wasn't edited; nothing to do. Stopping.");
+            return Ok(1);
+        }
+        BuildRewordMessageResult::EmptyMessage => {
+            eprintln!("Error: the reworded message was empty. Stopping.");
+            return Ok(1);
+        }
+        BuildRewordMessageResult::Failed { message } => {
+            if message != None {
+                eprintln!("{}", message.unwrap());
+            }
+            eprintln!("Error: problem processing commit message. Nothing reworded.");
+            return Ok(1);
+        }
+    };
+
+    if commits.len() != 1
+        && !force
+        && !Confirm::new()
+            .with_prompt(format!(
+                "Warning: attempting to apply the same message to {} commits. Continue?",
+                commits.len()
+            ))
+            .default(false)
+            .interact()?
+    {
+        println!("Ok. Nothing reworded.");
+        return Ok(1);
+    }
+
+    let subtree_roots = find_subtree_roots(&repo, &dag, &commits)?;
+
+    // TODO instead of calling into `r#move` shoud we just be building the rebase plan ourselves?
+    // TODO look at `sync` to see how it handles return values for multiple subtree moves
+    // TODO what if there is no parent?
+    // TODO what if there are multiple parents?
+    let move_options = MoveOptions {
+        force_in_memory: rebase_options.force_in_memory,
+        force_on_disk: rebase_options.force_on_disk,
+        detect_duplicate_commits_via_patch_id: rebase_options.detect_duplicate_commits_via_patch_id,
+        resolve_merge_conflicts: rebase_options.resolve_merge_conflicts,
+        dump_rebase_constraints: rebase_options.dump_rebase_constraints,
+        dump_rebase_plan: rebase_options.dump_rebase_plan,
+    };
+    for subtree_root in subtree_roots.iter() {
+        let base_commit = subtree_root;
+        let base_parents = subtree_root.get_parents();
+        let parent = base_parents.get(0).unwrap();
+        r#move(
+            effects,
+            git_run_info,
+            Some(base_commit.get_oid().to_string()),
+            Some(parent.get_oid().to_string()),
+            None, // base
+            Some(messages.clone()),
+            &move_options,
+        )?;
+    }
```

Comment:
Yep, we'll want to build and execute the rebase plan ourselves. We don't want to call `move` directly because it prints a bunch of auxiliary output that won't be relevant for `git reword`. I think it will also make the `find_subtree_roots` call unnecessary. (For future reference, I think what you wanted was `dag.query().roots(commit_set)`?)

 It should look something like this:

```rust
let rebase_plan = {
    let mut builder = RebasePlanBuilder::new(...);
    for (commit_oid, message) in messages {
        let replacement_oid = repo.amend_commit(None, ...)?;
        builder.replace_commit(commit_oid, replacement_oid)?;
        // TODO: handle the case of not-exactly-one parent.
        builder.move_subtree(commit_oid, commit_oid.get_only_parent());
    }
};
execute_rebase_plan(...)?;
```

It should actually be pretty straightforward to make `move_subtree` take a variable number of parents, but let's skip that for now and just `eyre::bail!` if we get that situation. I can fix it later.

Reply:
Gah! So brilliantly simple! So much code I didn't need to write! (I'll just keep telling myself it was fun.) And I was so proud of my `find_subtree_roots` algorithm ... of course there's a ready made 1-liner to replace it! 😄  I'll work on cleaning this up tomorrow (or later this week.) Thanks again!

Reply:
Ah, yes, this is *much* nicer. I'm going to leave `find_subtree_roots` in place for the time being, but in a drastically simplified format that uses `dag.query().roots()`. I might inline that later, but for now it still feels like nice encapsulation.